crates/sui-core/src/validator_client_monitor/monitor.rs

```diff
@@ -0,0 +1,299 @@
+    /// 1. The top `k` validators by score (shuffled)
+    /// 2. The remaining validators ordered by score (not shuffled)
+    pub fn select_shuffled_preferred_validators(
+        &self,
+        committee: &Committee,
+        k: usize,
+    ) -> Vec<AuthorityName> {
+        // Get all validators with their stats and scores
+        // Drop the lock immediately after getting the data
+        let all_validator_stats = self.client_stats.read().get_all_validator_stats(committee);
+
+        // Convert to a vector for sorting, handling None summaries
+        let mut validators_with_info: Vec<_> = all_validator_stats
+            .into_iter()
+            .map(|(validator, summary_opt)| {
+                if let Some(summary) = summary_opt {
+                    // Update metrics for validators with scores
+                    let validator_str = validator.concise().to_string();
+                    self.metrics
+                        .consecutive_failures
+                        .with_label_values(&[&validator_str])
+                        .set(summary.consecutive_failures as i64);
+                    self.metrics
+                        .performance_score
+                        .with_label_values(&[&validator_str])
+                        .set(summary.score);
+
+                    (validator, Some((summary.score, summary.is_excluded)))
+                } else {
+                    // No score available (not in committee or missing data)
+                    (validator, None)
+                }
+            })
+            .collect();
+
+        // Sort by:
+        // 1. Has score (validators with scores first)
+        // 2. Exclusion status (non-excluded first)
+        // 3. Score (highest first)
+        validators_with_info.sort_by(|a, b| {
+            match (&a.1, &b.1) {
+                (Some((score_a, excluded_a)), Some((score_b, excluded_b))) => {
+                    // Both have scores - sort by exclusion status, then score
+                    match (excluded_a, excluded_b) {
+                        (false, true) => std::cmp::Ordering::Less,
+                        (true, false) => std::cmp::Ordering::Greater,
+                        _ => score_b.partial_cmp(score_a).unwrap(),
+                    }
+                }
+                (Some(_), None) => std::cmp::Ordering::Less, // a has score, b doesn't
+                (None, Some(_)) => std::cmp::Ordering::Greater, // a has no score, b does
+                (None, None) => std::cmp::Ordering::Equal,   // neither has score
+            }
+        });
+
+        let committee_validators: Vec<_> = validators_with_info
+            .into_iter()
+            .map(|(validator, _info)| validator)
+            .collect();
+
+        // Shuffle only the top k elements in place, leaving the rest sorted
+        let mut result = committee_validators;
+        let k = k.min(result.len());
+        let mut rng = rand::thread_rng();
+        result[..k].shuffle(&mut rng);
```

Comment:
`result[k..]` should shuffle as well because their ordering is deterministic.

Reply:
I intentionally left the `k..` to not shuffle because: in 99%+ of cases, top k should be sufficient to serve the request. If in the unfortunate case where top k is not, it does not help to pick a completely random validator because then the performance becomes too unpredictable. Instead, we should just pick starting from the best of the remaining. And because this should be rare, we should not need to worry about load concentration here.